lib/plugins/crowdsec/stream.lua

```diff
@@ -0,0 +1,257 @@
+
+  -- As this function is running inside coroutine (with ngx.timer.at),
+  -- we need to raise error instead of returning them
+
+  if api_url == "" then
+    return "No API URL defined"
+  end
+
+
+  set_refreshing(true)
+
+  local is_startup = stream.cache:get("startup")
+  ngx.log(ngx.INFO, "startup: " .. tostring(is_startup))
+  ngx.log(ngx.DEBUG, "Stream Query from worker : " .. tostring(ngx.worker.id()) .. " with startup "..tostring(is_startup) .. " | premature: ")
+  local link = api_url .. "/v1/decisions/stream?startup=" .. tostring(is_startup)
+  local res, err = utils.get_remediation_http_request(link,
+                                                      timeout,
+                                                      api_key_header,
+                                                      api_key,
+                                                      user_agent,
+                                                      ssl_verify)
+  if not res then
+    set_refreshing(false)
+    ngx.log(ngx.ERR, "request failed: " .. err)
+  end
+
+  local succ, err, forcible = stream.cache:set("last_refresh", ngx.time())
+  if not succ then
+    ngx.log(ngx.ERR, "Failed to set last_refresh key in cache: " .. err)
+  end
+  if forcible then
+    ngx.log(ngx.ERR, "Lua shared dict (crowdsec cache) is full, please increase dict size in config")
+  end
+
+  local status = res.status
+  local body = res.body
+
+  ngx.log(ngx.DEBUG, "Response:" .. tostring(status) .. " | " .. tostring(body))
+
+  if status~=200 then
+    set_refreshing(false)
+    ngx.log(ngx.ERR, "HTTP error while request to Local API '" .. status .. "' with message (" .. tostring(body) .. ")")
+  end
+
+  local decisions = cjson.decode(body)
+
+  -- process deleted decisions
+  local deleted = {}
+  if type(decisions.deleted) == "table" then
+    for _, decision in pairs(decisions.deleted) do
+      if decision.origin == "lists" and decisions.scenario ~= nil then
+        decision.origin = "lists:" .. decision.scenario
+      end
+
+      self:delete(utils.item_to_string(decision.value, decision.scope))
+      if decision.type == "captcha" then
+        stream.cache:delete("captcha_" .. decision.value)
+      end
+      local key,_ = utils.item_to_string(decision.value, decision.scope)
+      local cache_value = stream.cache:get(key)
+      if cache_value ~= nil then
+        stream.cache:delete(key)
+        if deleted[decision.origin] == nil then
+          deleted[decision.origin] = 1
+        else
+          deleted[decision.origin] = deleted[decision.origin] + 1
+        end
+      end
+      ngx.log(ngx.DEBUG, "Deleting '" .. key .. "'")
+    end
+  end
+
+  -- process new decisions
+  if type(decisions.new) == "table" then
+    for _, decision in pairs(decisions.new) do
+      if decision.origin == "lists" and decision.scenario ~= nil then
+        decision.origin = "lists:" .. decision.scenario
+      end
```

Comment:
while captcha doesn't seem to have a specific treatment here

Reply:
It's a different scope of cache; In this cache, it's used to know if there's a remediation linked to the ip.